### PR TITLE
placeholders: deny access to internals and other unspecified stuff

### PR DESCRIPTION
**Pull request by enkore**
Also better error messages.

`Invalid placeholder "horstname" in string: testrepo::{horstname}-{now}`

### CODE REVIEW

File: src/borg/archiver.py [hunk 1]
```diff
@@ -1755,6 +1755,10 @@ def do_break_lock(self, args, repository):
 
             The version of borg, only major, minor and patch version, e.g.: 1.0.8
 
+        If literal curly braces need to be used, double them for escaping::
+
+            borg create /path/to/repo::{{literal_text}}
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: hmm, does a (non-coding) reader understand "literal" and "escaping"?

File: src/borg/helpers.py [hunk 2]
```diff
@@ -780,8 +784,15 @@ def __format__(self, format_spec):
 
 
 def format_line(format, data):
+    for _, key, _, conversion in Formatter().parse(format):
+        if not key:
+            continue
+        if '.' in key or '__' in key or conversion:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: instead of blacklisting . and __ we could also use a whitelist.

File: src/borg/archiver.py (hunk 1 above)
Comment (enkore) [inline src/borg/archiver.py]: borg's a shell tool... shell escaping knowledge is required :(

File: src/borg/helpers.py [hunk 3]
```diff
@@ -780,8 +784,15 @@ def __format__(self, format_spec):
...
 def format_line(format, data):
+    for _, key, _, conversion in Formatter().parse(format):
+        if not key:
+            continue
+        if conversion or key not in data:
+            raise InvalidPlaceholder(key, format)
     try:
-        return format.format(**data)
+        return format.format_map(data)
+    except KeyError as ke:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: can this KeyError still happen considering 790/791?

**Pull request merged**